ユーザー管理機能の実装

RyOsUkE0730 opened the pull request:
## what
deviseを用いた新規登録、ログイン画面のview及びUsersモデルの作成

## why
ユーザー管理機能を実装するため

各挙動に関するGyazoファイルは以下からご確認をお願い致します。
・[必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画](https://gyazo.com/49949a28ff026b8622fbbc1463ef533d)
・[入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画](https://gyazo.com/4ddf9836383a15d2ec6bfb688f46b2ec)
・[必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
](https://gyazo.com/bec8765f0cd11bd728cdd3ceff3b6475)
・[入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画](https://gyazo.com/fedc257c1155514c1c8d4cfd4ac48d93)
・[ログアウトに成功する動画](https://gyazo.com/be5652956bac35f5d6f5d41849608d15)
・[テストに成功する画像](https://gyazo.com/1ac820d07eb7c1f6c615a13899e1fcc3)

RyOsUkE0730 committed:
Install devise

RyOsUkE0730 committed:
Create Users table

RyOsUkE0730 committed:
Placement of view/devise

RyOsUkE0730 committed:
Update devise/new

・form_withメソッドの更新
・その他属性値の更新(hoge)

RyOsUkE0730 committed:
Update devise/new ver.2

・エラー発生時のメッセージ表示

RyOsUkE0730 committed:
Settings Strong parameter

RyOsUkE0730 committed:
Update top/header path

・_headerのlink_toメソッドを更新
・ログイン有無で表示の分岐を設定

RyOsUkE0730 committed:
Update top/header path ver.2

・link_toメソッドを正しい構成に修正

RyOsUkE0730 committed:
Update session/new

・form_withメソッドの更新
・その他属性値の更新(hoge)

RyOsUkE0730 committed:
others

不要な記述の削除

RyOsUkE0730 committed:
Settings validation

・新規登録に関するvalidatesの設定

RyOsUkE0730 committed:
Preparing for the test

・rspec/FactoryBot/Fakerの導入
・UsersモデルのFactoryBotを指定

RyOsUkE0730 committed:
Create test code

・Usersモデルに紐づくテスココードの作成
※必須条件の、validatesのみ
name/email/password

RyOsUkE0730 committed:
Play Rubocop

RyOsUkE0730 committed:
不要なコメントアウトの削除

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,13 @@
+         :recoverable, :rememberable, :validatable
+
+  validates :nickname, presence: true
+  validates :first_name, presence: true
+  validates :last_name, presence: true
```
```ruby
  validates :first_name, presence: true
  validates :last_name, presence: true
```
ユーザー名は姓名共に、全角（漢字・ひらがな・カタカナ）での入力を要求するバリデーションも指定しましょう！

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,13 @@
+  validates :nickname, presence: true
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+  validates :first_name_kana, presence: true
+  validates :last_name_kana, presence: true
```
```ruby
  validates :first_name_kana, presence: true
  validates :last_name_kana, presence: true
```
ユーザー名のフリガナは姓名共に、全角（カタカナ）での入力を要求するバリデーションも指定しましょう！

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,13 @@
+  validates :last_name, presence: true
+  validates :first_name_kana, presence: true
+  validates :last_name_kana, presence: true
+  validates :birthday, presence: true
```
パスワードの半角英数字混合での入力を要求するバリデーションも追記しましょう！

minachanchan commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,14 @@
+  factory :user do
+    nickname { Faker::Name.name }
+    email { Faker::Internet.free_email }
+    password = Faker::Internet.password(min_length: 6)
```
```ruby
password = Faker::Internet.password(min_length: 6)
```

Fakerでのpasswordの生成について記述できていて素晴らしいです！
こちらの記述ですと、稀に英字のみ、もしくは数値のみの値が生成されることがあります。
その際、モデルファイルでコメントさせていただいた、英数字混合での入力を要求するバリデーションを通過できないことになってしまいます。
以下のように、文字列として英数字を予め含めておきましょう！

```
password =  '1a' + Faker::Internet.password(min_length: 6)
```

minachanchan commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        @user.password_confirmation = '222222'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password confirmation doesn't match Password")
+      end
```
passwordに関しては以下の項目について確認するテストコードをそれぞれ記述しましょう！

```
・英字のみでは登録できないこと
・数字のみでは登録できないこと
・全角だと登録できないこと
```

minachanchan commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        @user.email = 'hogehogehoge'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
```
①
ユーザー名に関しては、姓名共に以下の項目について確認するテストコードをそれぞれ記述しましょう！

```
・空では登録できないこと
・半角文字が含まれている場合は登録できないこと
```

②
ユーザー名フリガナに関しては、姓名共に以下の項目について確認するテストコードをそれぞれ記述しましょう！

```
・空では登録できないこと
・カタカナ以外の文字（ひらがなや漢字）が含まれている場合は登録できないこと
```

③
birthdayが空だと登録できないことを確認するテストコードを追記しましょう！

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。

修正の後、再度レビューのご依頼頂けますようお願い申し上げます。
その際、最新のテスト結果の画像も合わせて添付いただけますようお願いいたします。

RyOsUkE0730 committed:
Update validates+

・passwordで半角英数字混合での入力を要求に変更
・姓/名で全角（漢字、ひらがな、カタカナ）での入力を要求に変更
・セイ/メイで全角カタカナでの入力を要求に変更

RyOsUkE0730 committed:
Update Factory/users

・passwordを半角英数に対応

RyOsUkE0730 committed:
Update test code+

・テストコードの追加、更新
・バリデーションの修正

RyOsUkE0730 commented:
## what
ご指摘いただいた以下の点を修正し、再度ユーザー管理機能の実装を行いました。
・バリデーションの制限の追加
・テスト実施項目の追加

## why
furima-appにおけるユーザー管理機能実装のため

テストコードの挙動については、[こちら](https://gyazo.com/e3fe9dbd6ae74eeecce0f11f7464ceda)からご確認をお願い致します。

yuhi-di commented:
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

ユーザー管理機能の実装、お疲れ様でした。
引き続き、応援しております。
